modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/streaming/PagingProviderProducer.java

```diff
@@ -91,27 +101,58 @@ public int getSize() {
    * @return
    */
   private <R> R performWithConnection(Function<Object, R> function) {
+    Optional<MutableConfigurationStats> stats = getMutableConfigurationStats(executionContext);
+    RetryPolicyTemplate retryPolicy =
+        (RetryPolicyTemplate) executionContext.getRetryPolicyTemplate().orElseGet(NoRetryPolicyTemplate::new);
+    CompletableFuture<R> future = retryPolicy.applyPolicy(() -> completedFuture(withConnection(function)),
+                                                          e -> !isFirstPage && !delegate.useStickyConnections()
+                                                              && shouldRetry(e, executionContext),
+                                                          e -> {
+                                                          },
+                                                          e -> stats.ifPresent(s -> s.discountInflightOperation()),
+                                                          identity(),
+                                                          executionContext.getCurrentScheduler());
+    try {
+      return future.get();
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof RuntimeException) {
+        throw (RuntimeException) e.getCause();
+      }
+      throw new MuleRuntimeException(createStaticMessage(COULD_NOT_EXECUTE), e.getCause());
+    } catch (InterruptedException e) {
+      throw new MuleRuntimeException(createStaticMessage(COULD_NOT_EXECUTE), e);
+    }
+  }
+
+  private <R> R withConnection(Function<Object, R> function) {
     ConnectionSupplier connectionSupplier = getConnectionSupplier();
     Object connection = getConnection(connectionSupplier);
     try {
       R result = function.apply(connection);
       return result;
-    } catch (Exception exception) {
+    } catch (Exception caughtException) {
       if (isFirstPage) {
         safely(() -> delegate.close(connection), e -> LOGGER.debug("Found exception closing paging provider", e));
       }
-      extractConnectionException(exception).ifPresent(ex -> {
-        if (isTransactional()) {
-          executionContext.setVariable(DO_NOT_RETRY, "true");
-        }
-        connectionSupplier.invalidateConnection();
-      });
-      throw exception;
+      extractConnectionException(caughtException).ifPresent(e -> handleConnectionException(e, connectionSupplier));
+      throw caughtException;
     } finally {
       safely(connectionSupplier::close, e -> LOGGER.debug("Found exception closing the connection supplier", e));
     }
   }
 
+  private void handleConnectionException(ConnectionException connectionException, ConnectionSupplier connectionSupplier) {
+    Optional<String> exceptionConfigName =
+        executionContext.getConfiguration().map(config -> ((ConfigurationInstance) config).getName());
+    if (isTransactional()) {
+      connectionException.addInfo("wasTransactional", true);
```

Comment:
Define constant in org.mule.runtime.module.extension.internal.ExtensionProperties. Also the key should start with the proper prefix.